File: 8-10.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10 ** 5)


def solution(n, A, m, Q):
    psum = list([0] * n for _ in range(n))

    for q in Q:
        if q[0] == 1:
            do_add_query(psum, q[1], q[2], q[3], q[4], q[5])
        else:
            for r in range(n):
                for c in range(1, n):
                    psum[r][c] += psum[r][c - 1]

            for c in range(n):
                for r in range(1, n, ):
                    psum[r][c] += psum[r - 1][c]

            print(get_sum(A, q[1], q[2], q[3], q[4]) + \
                  get_sum(psum, q[1], q[2], q[3], q[4]))


def do_add_query(A, i1, j1, i2, j2, k):
    A[i1][j1] += k
    if j2 + 1 < n:
        A[i1][j2 + 1] -= k
    if i2 + 1 < n:
        A[i2 + 1][j1] -= k
    if i2 + 1 < n and j2 + 1 < n:
        A[i2 + 1][j2 + 1] += k


def get_sum(A, i1, j1, i2, j2):
    ret = 0
    for i in range(i1, i2 + 1):
        for j in range(j1, j2 + 1):
            ret += A[i][j]
    return ret


n, m = map(int, input().split())
A = list(list(map(int, input().split())) for _ in range(n))
Q = list(list(map(int, input().split())) for _ in range(m))
solution(n, A, m, Q)

File: 8-10.cpp
#include <iostream>
#include <vector>
using namespace std;

void do_add_query(vector<vector<long long>>& A, int i1, int j1, int i2, int j2, int k) {
    A[i1][j1] += k;
    if (j2 + 1 < A.size()) {
        A[i1][j2 + 1] -= k;
    }
    if (i2 + 1 < A.size()) {
        A[i2 + 1][j1] -= k;
    }
    if (i2 + 1 < A.size() && j2 + 1 < A.size()) {
        A[i2 + 1][j2 + 1] += k;
    }
}

long long get_sum(vector<vector<long long>>& A, int i1, int j1, int i2, int j2) {
    long long ret = 0;
    for (int i = i1; i <= i2; i++) {
        for (int j = j1; j <= j2; j++) {
            ret += A[i][j];
        }
    }
    return ret;
}

void solution(int n, vector<vector<long long>>& A, int m, vector<vector<int>>& Q) {
    vector<vector<long long>> psum(n, vector<long long>(n, 0));
    for (auto q : Q) {
        if (q[0] == 1) {
            do_add_query(psum, q[1], q[2], q[3], q[4], q[5]);
        }
        else {
            for (int r = 0; r < n; r++) {
                for (int c = 1; c < n; c++) {
                    psum[r][c] += psum[r][c - 1];
                }
            }
            for (int c = 0; c < n; c++) {
                for (int r = 1; r < n; r++) {
                    psum[r][c] += psum[r - 1][c];
                }
            }
            cout << get_sum(A, q[1], q[2], q[3], q[4]) + get_sum(psum, q[1], q[2], q[3], q[4]) << endl;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;
    vector<vector<long long>> A(n, vector<long long>(n, 0));
    vector<vector<int>> Q(m, vector<int>(6, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> A[i][j];
        }
    }
    for (int i = 0; i < m; i++) {
        cin >> Q[i][0];
        int t;
        if (Q[i][0] == 1) t = 6;
        else t = 5;
        for (int j = 1; j < t; j++) {
            cin >> Q[i][j];
        }
    }
    solution(n, A, m, Q);
    return 0;
}